pep-9999.rst

```diff
@@ -0,0 +1,736 @@
+
+The type checker should reject this program: all three uses of
+``Literal`` are *invalid* according to this spec. However, Python itself
+should execute this program with no errors.
+
+This is partly to help us preserve flexibility in case we want to expand the
+scope of what ``Literal`` can be used for in the future, and partly because
+it is not possible to detect all illegal parameters at runtime to begin with.
+For example, it is impossible to distinguish between ``Literal[1 + 2]`` and 
+``Literal[3]`` at runtime.
+
+Literals, enums, and forward references
+---------------------------------------
+
+One potential ambiguity is between literal strings and forward
+references to literal enum members. For example, suppose we have the
+type ``Literal["Color.RED"]``. Does this literal type
+contain a string literal or a forward reference to some ``Color.RED``
+enum member?
+
+In cases like these, we always assume the user meant to construct a
+literal string. If the user wants a forward reference, they must wrap
+the entire literal type in a string -- e.g. ``"Literal[Color.RED]"``.
+
+Literals, enums, and Any
+------------------------
+
+Another ambiguity is when the user attempts to use some expression that
+is meant to be an enum but is actually of type ``Any``. For example,
+suppose a user attempts to import an enum from a package with no type hints::
+
+   from typing import Literal
+   from lib_with_no_types import SomeEnum  # SomeEnum has type 'Any'!
+
+   # x has type `Literal[Any]` due to the bad import
+   x: Literal[SomeEnum.FOO]
+
+Because ``Literal`` may not be parameterized by ``Any``, this program
+is *illegal*: the type checker should report an error with the last line.
+
+In short, while ``Any`` may effectively be used as a placeholder for any
+arbitrary *type*, it is currently **not** allowed to serve as a placeholder
+for any arbitrary *value*.
+
+Type inference
+==============
+
+This section describes a few rules regarding type inference and
+literals, along with some examples.
+
+Backwards compatibility
+-----------------------
+
+When type checkers add support for Literal, it's important they do so
+in a way that preserves backwards-compatibility. Code that used to
+type check **must** continue to type check after support for Literal
+is added.
+
+This is particularly important when performing type inference. For
+example, given the statement ``x = "blue"``, should the inferred
+type of ``x`` be ``str`` or ``Literal["blue"]``?
+
+This PEP does not require any particular strategy for cases like this,
+apart from requiring that backwards compatibility is maintained.
+
+For example, one valid strategy might be to analyze the surrounding code
+to deduce the appropriate type for ``x``. Another valid strategy would
+be to assume that ``x`` is always of type ``str`` and require users to
+manually annotate it if they want it to have a Literal type. A third strategy
+might be to use a hybrid of the previous two.
```

Comment:
I think this paragraph is kind of confusing, and ultimately doesn't matter because it says that these are only some of many possible strategies.  I think the section would be more readable if it were just removed.

Reply:
I'm not sure I feel comfortable with removing this paragraph entirely -- although the exact strategy for performing inference is now left unspecified, I do think it'll be helpful for readers if they see some examples of what type checkers are likely to do here in practice.

In particular, I want to make sure readers don't leave with the impression that type checkers are necessarily obligated to do something complicated here, and I think mentioning the "it's fine to do nothing new" strategy would help set user expectations.

I do agree this paragraph is worded poorly though. I went ahead and rewrote it -- does the new version seem better to you?